Update user table

tsungtingdu opened the pull request:
- Upgrade ruby to 2.4.2, in order to solve the problem of **Could not find "gem" in any of the sources** when generating model, migration, controller, etc..
- Add a new column "name" in users table

tsungtingdu committed:
Set up application with:

- 7 models
- Devise
- Bootstrap

tsungtingdu committed:
upgrade ruby to 2.4.2

tsungtingdu committed:
add a new column "name" to users table

tsungtingdu committed:
add dependent option

gohdaniel15 commented on config/environments/production.rb:
```diff
@@ -88,4 +88,5 @@
 
   # Do not dump schema after migrations.
   config.active_record.dump_schema_after_migration = false
+   config.action_mailer.default_url_options = { :host => 'localhost:3000' }
```
Hmmm, this shouldn't be here.

gohdaniel15 commented on config/routes.rb:
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
+  devise_for :users
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+root "instructors#index"
```
We're missing an indentation here. Also, as an application convention, lets stick to single-quotes `'` unless we're using string interpolation (e.g: `"#{ruby-code-here}"`)

gohdaniel15 commented on db/migrate/20171113153142_create_instructors.rb:
```diff
@@ -0,0 +1,9 @@
+    create_table :instructors do |t|
+    	t.string :name
+    	t.string :email
+      t.timestamps
```
There is some weird indentation going on here 😅

gohdaniel15 replied:
Also, name and email should have database null-restrictions, to ensure that our database has its own protections against invalid data.
```
  t.string :name, null: false
  t.string :email, null: false
```

tsungtingdu replied:
Regarding the indentation, there is no problem in my local document...weird

tsungtingdu replied:
![Uploading Screenshot 2017-11-20 22.01.42.png…](Image)

gohdaniel15 commented on db/migrate/20171113153157_create_students.rb:
```diff
@@ -0,0 +1,10 @@
+    	t.string :name
+    	t.string :email
+    	t.belongs_to :batch, index: true
+      t.timestamps
```
Weird indentation here too

gohdaniel15 replied:
Also, `name` and `email` should have database null-restrictions, to ensure that our database has its own protections against invalid data. 

```
  t.string :name, null: false
  t.string :email, null: false
```

gohdaniel15 commented on db/migrate/20171113153218_create_challenges.rb:
```diff
@@ -0,0 +1,10 @@
+    	t.string :title
+    	t.string :description
+    	t.integer :score
+      t.timestamps
```
Same here, for indentations and null-restrictions

gohdaniel15 commented on db/migrate/20171113153232_create_batches.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateBatches < ActiveRecord::Migration[5.1]
+  def change
+    create_table :batches do |t|
+    	t.string :track
```
What will track represent? Like `Web Development`, `Digital Marketing` etc? If so, then it seems like using an integer value here, and adding a `enum` on the `Batch` model would be more appropriate for this.

gohdaniel15 commented on db/migrate/20171113153232_create_batches.rb:
```diff
@@ -0,0 +1,11 @@
+    	t.string :track
+    	t.string :serial_number
+    	t.date :start_date
+    	t.date :end_date
```
All of these needs null-restrictions too :)

gohdaniel15 commented on db/migrate/20171113153954_create_transactions.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateTransactions < ActiveRecord::Migration[5.1]
+  def change
+    create_table :transactions do |t|
+    	t.date :transaction_date
```
we don't need this field, since the `:created_at` field generated by AR will already track this for us 😅

gohdaniel15 commented on db/migrate/20171113153954_create_transactions.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :transactions do |t|
+    	t.date :transaction_date
+    	t.integer :transaction_score
```
I think the prefix `:transaction_` here is unnecessary, since calling it would look something like `@transaction.transaction_score`. It would read better if it was just `@transaction.score` IMO

gohdaniel15 commented on db/migrate/20171113154028_create_challenge_acceptances.rb:
```diff
@@ -0,0 +1,10 @@
+    create_table :challenge_acceptances do |t|
+      t.belongs_to :challenge ,index: true
+      t.belongs_to :student, index: :true
+      t.string :status
```
`status` here should be an integer, mapped to an `enum` in the model too :)

gohdaniel15 commented on db/migrate/20171116151437_add_column_to_user.rb:
```diff
@@ -0,0 +1,13 @@
+  def down
+    raise ActiveRecord::IrreversibleMigration
+  end
+end
```
Since we're still so early in development, I would suggest editing the `DeviseCreateUsers` migration and dropping + creating + migrating the whole database again. We can't do that when we're in production and have live data coming in, but right now this is fine. The upside of doing this is that our migration's can serve as a record for any developer coming into this app, who will be able to easily see all the columns generated for this table with ease, instead of having to open each migration up to see the paper-trail of migrations that eventually form the table.

gohdaniel15 commented on app/models/student.rb:
```diff
@@ -0,0 +1,9 @@
+  has_many :challenges, through: :transactions
+  belongs_to :batch
+  has_many :chllenge_acceptances, dependent: :destroy
+  has_many :challenges, through: :chllenge_acceptances
```
Lets separate the `has_many` and `belongs_to` for easier readability. 
```
+  has_many :transactions, dependent: :destroy
+  has_many :challenges, through: :transactions
+  has_many :chllenge_acceptances, dependent: :destroy
+  has_many :challenges, through: :chllenge_acceptances

+  belongs_to :batch
```

gohdaniel15 commented on app/models/student.rb:
```diff
@@ -0,0 +1,9 @@
+  has_many :challenges, through: :transactions
+  belongs_to :batch
+  has_many :chllenge_acceptances, dependent: :destroy
+  has_many :challenges, through: :chllenge_acceptances
```
Also, there's something wrong here. A student should not be able to have two `has_many :challenges`. If we think about it, when we call `@student.challenges`, then which one will rails resolve to?

gohdaniel15 replied:
Also, there appears to be a typo for the `:chllenge_acceptances` table? 😅 Lets fix that first

gohdaniel15 reviewed (changes requested):
Good work @tsungtingdu! I'm being very nitpicky here, since these set up the foundation for the rest of the app, and I want to get it right from the start so it won't be a mess for us to deal with later on. 

Try to work on as many of these changes as you can first, and I'll cover the rest of the explanations on Monday at work 👍

tsungtingdu committed:
- Seperate has_many nad belongs_to to improve readability in studnet.rb
- Delete config.action_mailer.defaul_url_options in environments/production.rb
- Fix the indentation and use single-quotes in routes.rb
- Generate a migration to amend tables but did not migrate yet

tsungtingdu committed:
migrate amend_tables

tsungtingdu committed:
Modify all migration files

tsungtingdu committed:
Tidy up codebase

gohdaniel15 reviewed (approved):
LGTM! 🌴
